### ISSUE TITLE
Support rendering in multiple CRS

### ISSUE THREAD

**Issue by jailln** [maplibre/maplibre-gl-js#168]
Many GIS need to have the ability to render maps in different coordinate reference systems, e.g. local ones, whether for precision or for displaying data of users in their source coordinate systems.
From what I know, maplibre always displays maps using the web mercator projection (EPSG:3857). 

Since I'm not familiar with maplibre code, I'm wondering how hard would it be to allow displaying maps in different projection systems ? Would it require a major refactoring ? 

In addition, I'm also wondering if other people would be interested by this feature and if it could be something that goes in the roadmap ?

Cheers

**Comment by wipfli** [comment]
As far as I know web mercator is the only projection available. And it sounded like it suits peoples needs quite well. You can transform mouse positions and marker locations on-the-fly I guess. Or what would be your use case for anything different than web mercator?

**Comment by kylebarron** [comment]
Other projections would be really nice, for example to work with polar data or to avoid web mercator distortion issues, but it would be a **HUGE** undertaking for a general case. First you'd need to support the rendering and placement of non-web mercator tiles. Then you'd need to create your own basemap because the vector tile basemap pipeline Mapbox uses only supports Web Mercator data.

Note that you can _hack_ alternative projections into Mapbox/Maplibre GL in specific circumstances. For example, the New York Times provides its interactive maps of the U.S. (e.g. elections and Covid cases) in an Equal Albers projection. (Note the northern border with Canada curves instead of being a straight line as it would be in Mercator).

![image](https://user-images.githubusercontent.com/15164633/121248197-61504c00-c860-11eb-97c6-0f563d1027ec.png)

If you're willing to do your own math for your custom vector tiles, you can achieve something like this by the lat/lon boundaries of `[-180, -90, 180, 90]` as an arbitrary canvas. So you'd convert your data from their coordinates in their non-Web Mercator projection to their relative points in this canvas _without reprojecting_. But of course those NYT maps can't use a Mapbox basemap because the Northeast of the U.S. would correspond to "upper right" in web mercator and load basemap data from Russia.

**Comment by davenquinn** [comment]
This seems quite out of scope for this library, as it basically only speaks Web Mercator. Other mapping libraries (ex. d3 as @kylebarron mentioned) are probably more fit for your needs here. There are also ways to "fool" web mercator for polar data etc.

**Comment by jailln** [comment]
> As far as I know web mercator is the only projection available. And it sounded like it suits peoples needs quite well. You can transform mouse positions and marker locations on-the-fly I guess. Or what would be your use case for anything different than web mercator?

Mainly to avoid web mercator distortions issues for e.g. measurements or edition, without having to transform positions on the fly every time. Also, to display raster data in local projections.


> Other projections would be really nice, for example to work with polar data or to avoid web mercator distortion issues, but it would be a **HUGE** undertaking for a general case. First you'd need to support the rendering and placement of non-web mercator tiles. Then you'd need to create your own basemap because the vector tile basemap pipeline Mapbox uses only supports Web Mercator data.
> 
> Note that you can _hack_ alternative projections into Mapbox/Maplibre GL in specific circumstances. For example, the New York Times provides [its interactive maps](https://www.nytimes.com/interactive/2021/us/covid-cases.html) of the U.S. (e.g. elections and Covid cases) in an Equal Albers projection. (Note the northern border with Canada curves instead of being a straight line as it would be in Mercator).
> 
> ![image](https://user-images.githubusercontent.com/15164633/121248197-61504c00-c860-11eb-97c6-0f563d1027ec.png)
> 
> If you're willing to do your own math for your custom vector tiles, you can achieve something like this by the lat/lon boundaries of `[-180, -90, 180, 90]` as an arbitrary canvas. So you'd convert your data from their coordinates in their non-Web Mercator projection to their relative points in this canvas _without reprojecting_. But of course those NYT maps can't use a Mapbox basemap because the Northeast of the U.S. would correspond to "upper right" in web mercator and load basemap data from Russia.

Thanks for this detailed answer! I guess that some parts of the code assume that tiles are are web-mercator and so th'at why it would require many changes to make it a general case.

> This seems quite out of scope for this library, as it basically only speaks Web Mercator. Other mapping libraries (ex. d3 as @kylebarron mentioned) are probably more fit for your needs here. There are also ways to "fool" web mercator for polar data etc.

I'm suprised that you consider this out of scope, could you elaborate on why please ? 


Also, I just found out that the mapbox team have started working on this: https://github.com/mapbox/mapbox-gl-js/issues/3184#issuecomment-831385021

**Comment by iacopoff** [comment]
TBH this would be extremely useful for rendering rasters representing graph-like data such as rivers. 
You don't want to reproject/regrid those rasters to Pseudo-Mercator as you would completely disrupt the topology.  
Is there an interest in moving this forward?

**Comment by wipfli** [comment]
https://roadmap.maplibre.org/c/91-custom-coordinate-system-epsg-non-mercator-projection

**Comment by erikvullings** [comment]
> As far as I know web mercator is the only projection available. And it sounded like it suits peoples needs quite well. You can transform mouse positions and marker locations on-the-fly I guess. Or what would be your use case for anything different than web mercator?

Another use case is the following: In The Netherlands, the government offers a free [vector tile service](https://app.pdok.nl/vectortile-viewer/) of the whole country, which you can include in your own web apps, However, they offer the vector tiles in RD/[EPSG:28992](https://epsg.io/28992), basically a rectangular grid introduced by Napoleon with the center in Paris. The reason is that many open data in NL is in this format too. However, I cannot use them in MapLibre and therefore need to host my own vector tiles, which is a nuisance.

**Comment by HarelM** [comment]
@erikvullings You might be able to use `addProtocol` to convert the data from one projection to another, but I'm not sure if the tiles use the regular x-y-z grid that is used by maplibre, and in that case I'm not sure what your options are.
If this something you would like to introduce to maplibre you are more than welcome to submit a PR :-)

**Comment by geekdenz** [comment]
A hint where to find the code with which tiles are read into geometries would be very helpful.

**Comment by HarelM** [comment]
The worker that reads the data from the network:
https://github.com/maplibre/maplibre-gl-js/blob/bd37870b6fcc613a021c523d4eb5aa8638d71382/src/source/vector_tile_worker_source.ts#L134C9-L134C9
Worker tile is responsible for parsing the data:
https://github.com/maplibre/maplibre-gl-js/blob/bd37870b6fcc613a021c523d4eb5aa8638d71382/src/source/worker_tile.ts#L89C19-L89C31
https://github.com/maplibre/maplibre-gl-js/blob/bd37870b6fcc613a021c523d4eb5aa8638d71382/src/source/worker_tile.ts#L102

This is just the parsing of the vector tiles, if you want pointer to other part of the code, let me know and I'll look them for you.
Having this in this library would be awesome, I'd love to help anyone who would like to help push this through.

**Comment by geekdenz** [comment]
Wow that was a super quick answer. Considering there is proj4js reprojecting on the fly might be reasonably doable. There might be some caveats though that I am missing. As always the devil is probably in the detail. But I am using this library for a project of mine and I need support for other projections.

**Comment by HarelM** [comment]
Let me know if you would like me to assign a bounty for it and which bounty size you think is appropriate.
Bounty direction:
https://github.com/maplibre/maplibre/issues/272

In order to avoid bounty size misalignment, I would consider creating a design first detailing where this change impacts as a first bounty and then decide on the implementation bounty size.

But this is just my thoughts, let me know how you would like to proceed.

**Comment by geekdenz** [comment]
I'll have a poke around and get a debug session going for a while before committing to anything. I might be biting off more than I can chew. However, this would be a project right up my alley and expertise.

**Comment by HarelM** [comment]
Cool, looking forward to hearing more from you.
Note that there are multiple `README.md` files scattered in this repo, each one with its own info, I would advise to try and read as many of them as you need.

**Comment by geekdenz** [comment]
Thanks @HarelM ! Definitely very interested, also in the native implementation since I did quite a bit of competitive programming with C++ and I've got extensive experience with TypeScript. Would be interested in the bounty as well as the work.

However, I'll need quite some time to familiarize myself with the project first. Planning on reading the native book and some source code before I drive into a design. If someone else beats me to it I'd encourage them. Maybe it could be a friendly competition as is the spirit of FOSS.

**Comment by geekdenz** [comment]
Would this need a styles change or would projections be a part of the API and handing it to the Map options?

If it is in styles, this might need quite a bit of thought and maybe discussion as hinted in the CONTRIBUTING.md file.

I guess you could provide optional projections in the style document and specify a default. But if we are reprojecting on the fly, why not just have it part of the API and Map options for now. That would be a faster way to get to the smaller goal.

What do you think?

**Comment by HarelM** [comment]
I think we can safely start with the API. After that (or in parallel) we can have a discussion about how it should be in the style.
I think the end goal would be to define it in the style, but a solution in the API would be a very good way to test it and see how it looks and behaves.

**Comment by geekdenz** [comment]
Apologies. Haven't replied for quite a while.

This is not the highest on my priority list at the moment. So I'm happy if someone else takes over.

However, I'm still very interested in this problem and apart from work and my current side project and of course family, this is the next item on my priority list. I expect to finish my first iteration on my side project in about 4 weeks which is when I'd need the projection feature of maplibre gl js.

**Comment by eslindsey** [comment]
I'd love to experiment with the performance of this library vs. Leaflet, but my application is for a game and so currently uses Leaflet's `CRS.Simple`. Support for non-Earthlike CRS would be fantastic!

**Comment by geekdenz** [comment]
I am still keen to work on this. I am time-constrained as an engaged father who also has a full time job and many commitments though so please understand that I cannot sink too much time into this.

Also, I am concerned it might be quite hard to achieve. But that is also the attraction. Plus it would help a hobby project of mine.

Would it make sense to do the 2D part of this first, since it is easier?

Is it as simple as using the proj4js library and hooking that in somewhere?

**Comment by HarelM** [comment]
I think an initial work on this has started as part of https://github.com/maplibre/maplibre-gl-js/issues/307 and https://github.com/maplibre/maplibre/discussions/161
2D only is a possible way, I think the projection of the vector stuff might be complicated and I'm not sure proj4js would solve everything easily, as if you only project back to lat-lon you still have issues with the poles, which is the main reason as far as I understand, people need other projections.
But I'm no export on this, below are the relevant people.
cc: @Pheonor @kubapelc

**Comment by kubapelc** [comment]
Hi, first of all, I've only skimmed through the discussion, so I might have missed some context. I'm implementing vector globe projection, which is a similar problem. I'm basically doing reprojection on-the-fly in the vertex shader, taking vertices in tile coordinates (basically mercator) and projecting them to a sphere. In my code, both old mercator projection and new globe projection use the same vertex buffer.

There is an important caveat though - since I'm only reprojecting the vertices of a polygon, its edges remain straight, which is a problem especially for large polygons. What is a straight edge in mercator projection will become a curve on a globe. To fix this, I first subdivide every polygon (and other features too), so that I get a better approximation of a curve when reprojecting the vertices. This subdivision step is relatively complex and it's hard to make it fast enough (otherwise tile loading gets noticeably slower).

Since proj4js seems to only reproject individual coordinates, the problem of edges not getting curved would remain. The easiest way to implement other projections would be to use my vector globe which already subdivides polygons and edges, and "just" replace the mercator->globe projection in the vertex shader with a custom projection, and possibly tweak how much subdivision happens at what zoom level. This would also mean that the reprojection code would need to be rewritten to GLSL.

**Comment by geekdenz** [comment]
Cool @kubapelc !

Thanks for reaching out.

I think reprojection can assume that only lines exist. The line segments should be small enough that curvature does not need to be implemented or if, it should be a setting that is off by default, especially if there is a performance hit.

But that is just my opinion and a thought.

Think agile and small increments. Curving lines is definitely a nice to have rather than a must imho.

**Comment by Pheonor** [comment]
Effectively, 'globe' could be consider as a specific projection and to switch between 'mercator' and 'globe' a first step on a projection class have been implemented but with very limited support.
This class should at least to encapsulate the project / unproject methods instead of direct call to `mercatorXfromLng`, `mercatorYfromLat`, `lngFromMercatorX` and `latFromMercatorY`.

My current version is based on an abstract class:
```
import {LngLat} from './lng_lat';

/**
 * A `ProjectedPoint` represents a projected point.
 * It allows to store the 3D coordinates of any point.
 */
export class ProjectedPoint {
    x: number;
    y: number;
    z: number;
}

/**
 * A `Projection` abstract class to represent a projection coordinate system.
 * It could represent a 3D globe projection or any 2D projection.
 * It allows to project and unproject coordinate to and from the coordinate system.
 *
 * @group Geography and Geometry
 */
export abstract class Projection {
    name: string;

    constructor(name: string) {
        this.name = name;
    }

    abstract isGlobe(zoom: number): boolean;
    getFactor(_zoom: number): number { return 0.0; }
    abstract project(lng: number, lat: number, zoom: number) : ProjectedPoint;
    abstract unproject(x: number, y: number, zoom: number) : LngLat;
}
```
But propably we should add some `options` into constructor to manage projection parameters.

As @kubapelc and @geekdenz explain, the tesselation of mesh to a sufficient level coud be sufficient to transform the mercator geometry into another projection and manage the different curvature.

**Comment by exotfboy** [comment]
Regarding multiple CRS, there can be varying interpretations:

- Vector tiles are created using diverse CRS, but Maplibre simply renders them as-is, specifying the CRS of the tile either through the API or in the style.

- Vector tiles are produced with one or more specific CRS, and Maplibre has the capability to project them dynamically.

I believe that the first option (1) is sufficient.

**Comment by wipfli** [comment]
@pka you did something like this, right?

**Comment by pka** [comment]
> @pka you did something like this, right?

I did option (1) to display tiles in Equal Earth projection with MapLibre:

![image](https://github.com/maplibre/maplibre-gl-js/assets/19007/47dd176a-5179-4474-8d74-f273878eee2f)

([Screencast](https://mapstodon.space/@implgeo/111992942539509582))

I'm currently investigating to make the tile grid more similar to Web Mercator, but for many applications the coordinate transformation from geographic coordinates to the tile CRS is needed. I plan to provide that for Equal Earth projection for all major map viewer libraries.

**Comment by HarelM** [comment]
@pka note that there's a globe effort which touches some aspects of this I believe.
The first PR is underway here: #3783.
I would recommend seeing which part is similar and what can be used to maybe create a PR that only addresses this part without the full implementation of a globe so that this can be added regardless of the globe's progress.
Meaning if we can have this feature implemented and merged while continue working on the globe in parallel, this way we bring more value to customers early on.

**Comment by scaddenp** [comment]
I am all for the simple version. Two use case - one is large no. of vector tile sets being published in our national grid system. They are the common and natural basemaps to use. No. 2 is for work in Antarctica. Web mercator is seriously misleading there.

**Comment by NicholasEwing** [comment]
Still waiting on this to be resolved - is there still no equivalent of Leaflet's `CRS.Simple`?

**Comment by geekdenz** [comment]
There might be a way to do this in an extension. Is this correct @HarelM ?

**Comment by HarelM** [comment]
Depending on what you would like to achieve. In theory, you could create a source that traslates to wgs84 (or via addProtocol), but I'm not sure this is the definition of supporting different CRSs...

**Comment by scaddenp** [comment]
That wouldn't work for displaying vector tiles in CRS of choice. Option 1 would be an improvement for us on this  - generally want to display vector tiles in the CRS they were created in. I would be keen to know how @pka achieved that. Was this on a fork of maplibre??

**Comment by dzfranklin** [comment]
For anyone looking to reproject raster tiles sticking openlayers in a custom protocol kinda works surprisingly well.

This example doesn't handle cancellation and the tile fetching is completely separate so it will have separate concurrency limits and so on.

https://jsfiddle.net/nfq6uLe1/35/

<details>
<summary><code>ml.addProtocol(...</code></summary>

```typescript
ml.addProtocol('os-explorer', async (params, abortController) => {
  const url = new URL(params.url);
  const path = url.pathname.replace(/^\/\//, '');

  if (path === 'tile') {
    const z = parseInt(url.searchParams.get('z')!);
    const x = parseInt(url.searchParams.get('x')!);
    const y = parseInt(url.searchParams.get('y')!);
    if (isNaN(z) || isNaN(x) || isNaN(y)) throw new Error('bad params');

    const data = await reprojectTile([z, x, y]);
    return { data };
  } else {
    throw new Error('not implemented: ' + params.url);
  }
});

const tileImageSource = new TileImageSource({
  url: 'https://api.os.uk/maps/raster/v1/zxy/Leisure_27700/{z}/{x}/{y}.png?key=' + OS_KEY,
  projection: 'EPSG:27700',
  wrapX: false,
  crossOrigin: '',
  tileGrid: new TileGrid({
    resolutions: [896.0, 448.0, 224.0, 112.0, 56.0, 28.0, 14.0, 7.0, 3.5, 1.75],
    origin: [-238375.0, 1376256.0]
  }),
  reprojectionErrorThreshold: errorThreshold,
});

function reprojectTile(coord: TileCoord): Promise<ArrayBuffer> {
  console.log(`reprojectTile: requesting ${coord.join('/')}`);
  return new Promise((resolve, reject) => {
    const tile = tileImageSource.getTile(
      coord[0],
      coord[1],
      coord[2],
      1,
      proj3857,
    ) as ReprojTile;

    tile.addEventListener('change', () => {
      switch (tile.getState()) {
        case TileState.LOADED: {
          const canvas = tile.getImage();
          canvas.toBlob((blob) => blob!.arrayBuffer().then(resolve));
          console.log(`reprojectTile: reprojected ${coord.join('/')}`);
          return;
        }
        case TileState.ERROR: {
          reject(new Error('openlayers: reprojection tile in error state'));
          return;
        }
      }
    });

    tile.load();
  });
}
```
</details>

**Comment by neodescis** [comment]
That's a really nifty idea. Thanks for sharing!

**Comment by birkskyum** [comment]
Since we have a globe implemented now, that solves the issue of mercator distortion, also in polar regions. If there is further need for custom 2D projections, feel free to open a new ticket for that, so we can better asses the demand.

**Comment by scaddenp** [comment]
As requested, I have created new ticket #5764

**Comment by willcohen** [comment]
As also noted at https://github.com/maplibre/maplibre/discussions/163#discussioncomment-16041740, a library to transparently transform a mapbox style using only geojson sources to web Mercator coords that appear reprojected (same hack as noted above) is available at https://github.com/willcohen/backproj (demo at https://willcohen.github.io/backproj). MVTs are likely achievable using the same core technique but will require more stitching and reslicing, so aren’t in a first release.

**Comment by johncarmack1984** [comment]
I'd like to take this on. I have a working branch that closes it, and a plan to land it in pieces.

This and #5764 are the same problem underneath: NZ tiles in EPSG:2193, sea ice in polar stereographic, `CRS.Simple` grids for people who don't want lnglat at all. The tiles come pre-projected, so the map doesn't reproject anything. It just has to stop assuming the world is Mercator. Today that assumption sits in the camera, tile cover, bounds, GeoJSON tiling, terrain, and hillshade.

The GPU reprojection lane (maplibre/maplibre-style-spec#1623, https://github.com/maplibre/maplibre/discussions/491) is a different problem: re-rendering Mercator tiles into Equal Earth and friends. The seam below is groundwork it can use too.

The API:

```ts
maplibregl.addProjection({
    name: 'EPSG:2193',
    project: (lng, lat) => proj4('EPSG:4326', 'EPSG:2193', [lng, lat]),
    unproject: (x, y) => proj4('EPSG:2193', 'EPSG:4326', [x, y]),
    tileMatrix: {origin: [-3260586.7284, 10438190.1652], extentAtZoom0: 10018754.1714}, // NZTM2000Quad
});
```

then `"projection": {"type": "EPSG:2193"}` in the style. Same shape as `addProtocol`. The spec already accepts any string for `projection.type`, so this costs it nothing. Users bring their own `project`/`unproject`, so proj4js never touches the bundle. A planar quad grid is affine per tile, so it rides the existing mercator shader variant: no new GLSL. `mercator` keeps its fast path. A built-in `simple` (identity) CRS covers the `CRS.Simple` ask and doubles as the test fixture.

Fenced on purpose: square power-of-two quad grids only (OGC 2D-TMS quad). No world copies, no antimeridian wrap, `renderWorldCopies` forced off. No globe transition; a CRS map stays flat. One CRS per map.

Internally it's one seam, `Projection.worldCoordinateHelper`: a `worldFromLngLat(lng, lat) → {x, y}` / `lngLatFromWorld` pair (not separable per axis, transverse Mercator and stereographic need both inputs), meters per world unit, and a `wraps` flag, plus `Projection.isPlanar` in place of the `name === 'mercator'` checks. Mercator's implementation is the old functions, pinned bit-identical against `main` by a test.

Nobody wants a 3,000-line PR, so, in slices, each green on its own:

1. Route lng/lat↔world math through the seam: transform, camera (`fitBounds`), tile cover. Behavior-neutral, no new API.
2. `addProjection`/`removeProjection`, `PlanarProjection`, the `simple` CRS, render tests. The API lands here.
3. Honor the map CRS in sources, `queryRenderedFeatures`, GeoJSON (pre-projected on the main thread, geojson-vt untouched), terrain, hillshade.
4. A `{bbox}` URL token in CRS units for WMS/WMTS-style endpoints.
5. Docs: a custom-CRS guide with the EPSG:2193 setup, a `simple` example, and a one-line style-spec docs PR saying `projection.type` accepts registered names. If people later want the CRS declared in the style itself (proj string plus tile matrix, no code), that's its own spec conversation, and this leaves room for it.

@HarelM, per your 2023 note: API first, style later. PR 1 is up: #8236.

**Comment by HarelM** [comment]
This would be a huge win if this lands, thanks for looking into it! 
Can you share a demo of the working branch so people can play with it? 
It might be useful to present a design first before rushing to implemtation. 
There's also some work started in maptiler to migrate some of the proj4 code into the GPU/shaders so you might want to sync with them.
I think thier main pain was around gestures, map movement etc, I don't know if you included that in your plan, but worth making sure you take that into consideration. 
Overall, very exciting news!

**Comment by johncarmack1984** [comment]
Thanks! All four, in order.

**Demo:** https://johncarmack1984.github.io/maplibre-gl-js/crs-demo/. Three panels: the built-in `simple` grid, NASA GIBS sea ice in EPSG:3413 through the branch's `{bbox}` token, and NZTM2000 (bring a free LINZ key, the page tells you where). Nothing is reprojected on the client. The camera lives in the CRS, so drag off the edge of the polar square and the map stops where the world ends.

**Gestures:** that's the first thing I checked, because it's where this usually dies. Drag, scroll, rotate, pitch, inertia, `easeTo`/`flyTo`, `fitBounds`: none of them touch projection math directly. They go through `setLocationAtPoint`, `screenPointToLocation` and the camera helper's pan interpolation, and all of those now resolve through one seam (below). Terrain picking goes through the same seam. The design has a section on it.

**Design first:** agreed. The comment above is the shape (API, seam, fences, slices); below is the rest of it, the parts that decide whether it works. #8236 is the behavior-neutral seam and can wait on this conversation.

**MapTiler:** I found it: @klokan's https://github.com/maplibre/maplibre/discussions/491, @Sorceror's maplibre/maplibre-style-spec#1623, and the Equal Earth demo at https://labs.maptiler.com/equal-earth-web-maps/. That's shader-side Equal Earth and Plate Carrée with an adaptive transition back to Mercator, and a camera helper rebuilt around a moving central meridian, which I'd guess is the gesture pain you mean. Different problem from this one (theirs reprojects Mercator tiles, mine draws pre-projected ones), same seam underneath: both need the camera, tile cover and bounds to stop assuming Mercator. I'll post on maplibre/maplibre#491 and ask them to look at the seam before they build their own.

---

## Design, the rest of it

### The full type

The comment showed the call; this is the whole definition. A second registration under the same name throws, like `addSourceType`.

```ts
type CrsDefinition = {
    name: string;                               // the value of projection.type in the style
    project(lng: number, lat: number): [number, number];
    unproject(x: number, y: number): [number, number];
    tileMatrix: {origin: [number, number]; extentAtZoom0: number}; // top-left of tile 0/0/0 and its width, in CRS units
};
```

World coordinates are the unit square tile 0/0/0 covers; zoom z is its 2^z split. That is already what the code means by "world". `simple` is origin [-90, 90], extent 180.

### Gestures and camera

Handlers never see projection math. `DragPan`, `ScrollZoom`, `DragRotate`/pitch and touch work in screen space and call `transform.setLocationAtPoint` / `screenPointToLocation` / zoom deltas; inertia and `easeTo`/`flyTo` interpolate in world coordinates inside the camera helper; `fitBounds` projects the box corners. All of that goes through the seam, so a gesture on a CRS map is the same code as on Mercator with different numbers.

What changes for a world that does not wrap: no antimeridian wrap, no world copies (`renderWorldCopies` reads `false` while a CRS is active; the setting survives a round trip back to Mercator), and constrain clamps to the world square, zooming in until it fills the screen. Terrain picking takes world coordinates from the same helper, so it is the #8213 path with the CRS numbers.

### Tiles and sources

Vector, raster and raster-dem tiles load unchanged; `{z}/{x}/{y}` is the tile matrix split. TileJSON `bounds` are honored in CRS space (the box is projected once, at source load). Image sources place their corners through the seam. `{bbox}` is minx,miny,maxx,maxy in CRS units, in the order `project` returns; `{bbox-epsg-3857}` becomes the Mercator case of the same code.

### GeoJSON

geojson-vt tiles in Mercator on the worker and I'm not touching it. For a CRS map the main thread moves every position to the lng/lat whose Mercator projection is the CRS world position, sends that, and unprojects results back through the CRS. Two costs, documented rather than hidden: URL data is fetched on the main thread on CRS maps (the user's `project` closure can't cross the worker boundary, same reason `addProtocol` registers on the worker), and geometry filters (`within`) don't work there.

### Terrain, hillshade, queries

Terrain samples the DEM by world coordinates through the helper, rendered surface first and covering tiles as fallback, as on main. Hillshade's latitude term is zero for a planar CRS. `queryRenderedFeatures` / `querySourceFeatures` unproject through the CRS. A CRS map stays flat; `GlobeControl` returns to whatever flat projection it left.

### Cost

Mercator is bit-identical (pinned). Hot files benched against `main` on an idle machine: 0.96–1.03x, noise. Bundle: +1.7 KB min for the seam, about +6.5 KB for the whole series, measured at every commit; the worker shrinks 0.4 KB.

**Comment by scaddenp** [comment]
This is huge - thanks very much for this work!

**Comment by jo-chemla** [comment]
Huge work indeed, thanks for the great write-up and follow-up regarding the suggested Design/API surface, and nice seeing it is meant to cover both CRS for georeferenced datasets (like NASA GIBS sea ice in EPSG:3413, LINZ NZTM2000 in EPSG:2193), but also simple-grid `CRS.Simple` for local-datasets! Great to see that what's in screen-space like user-controls and camera stays there, and that most of the rest is computed in the honored CRS space (tiles/sources). 

I'm especially interested in testing the terrain/hillshading on DTM with custom CRS via [terrain-viewer](https://terrain-viewer.iconem.com/) ([repo](https://github.com/iconem/terrain-viewer)), where users can BYOD bring their own datasets (terrain or basemaps, tiles TMS, COG via `geomatico/maplibre-cog-protocol`, WMS etc). These BYOD spirces are not necessarily stored/expressed in EPSG:3857, so it would allow displaying such datasets in the dataset space

At some point, the maptiler design proposal linked above for the map itself to be reprojected GPU-side to eg equal-earth or other, would also be helpful, eg to even mix various dataset in different CRSes.
